ntp-daemon/src/server.rs

```diff
@@ -160,3 +154,86 @@ fn accept_packet(
         }
     }
 }
+
+#[derive(Debug)]
+struct TimestampedCache<T> {
+    elements: Vec<Option<(T, std::time::Instant)>>,
```

Comment:
All the qualified types look pretty unidiomatic to me, any reason for that?

Also, this could use some documentation on what the design is here. Is it a fixed-length compact hash map of some sort? What is it's purpose? What are we trying to optimize for?

Reply:
The planned use is in rate limiting: we keep track of when a peer last checked in. If it checks in too often, we issue a rate limiting KISS code.

A standard hashmap could become very large when we have many connections or when the server has been running for a long time (or you would have to garbage collect it periodically, which has its own complications). We thought about an adapter type with a custom hash instance (that takes the modulo), but this seemed simpler.

Reply:
Agreed, bit better documentation of that would be better.